contracts/metatx/ERC2771Forwarder.sol

```diff
@@ -0,0 +1,278 @@
+
+    /**
+     * @dev The request `deadline` has expired.
+     */
+    error ERC2771ForwarderExpiredRequest(uint48 deadline);
+
+    /**
+     * @dev See {EIP712-constructor}.
+     */
+    constructor(string memory name) EIP712(name, "1") {}
+
+    /**
+     * @dev Returns `true` if a request is valid for a provided `signature` at the current block.
+     *
+     * A transaction is considered valid when it hasn't expired (deadline is not met), and the signer
+     * matches the `from` parameter of the signed request.
+     *
+     * NOTE: A request may return false here but it optionally reverts in {batchExecute} to prevent
+     * reverts when a batch includes a request that was already frontrunned by another relay. This
+     * behavior is opt-in by using a flag in {executeBatch}.
+     */
+    function verify(ForwardRequestData calldata request) public view virtual returns (bool) {
+        (bool alive, bool signerMatch, ) = _validate(request, nonces(request.from));
+        return alive && signerMatch;
+    }
+
+    /**
+     * @dev Executes a `request` on behalf of `signature`'s signer guaranteeing that the forwarded call
+     * will receive the requested gas and no ETH is left stuck in the contract.
+     */
+    function execute(ForwardRequestData calldata request) public payable virtual returns (bool) {
+        // This check can be before the call because _execute reverts with an invalid request.
+        if (msg.value != request.value) {
+            revert ERC2771ForwarderMismatchedValue(request.value, msg.value);
+        }
+
+        return _execute(request, _useNonce(request.from), true);
+    }
+
+    /**
+     * @dev Batch version of {execute} with optional atomic behavior and refunding.
+     *
+     * The `atomic` parameter indicates whether all requests are executed or the whole
+     * batch reverts if a single one of them is not a valid request as defined by {verify}.
+     *
+     * The `refundReceiver` is used for optionally send the funds back to a selected address in
+     * case there's ETH left in the contract at the end of the execution instead of reverting. If
+     * the provided refund receiver is the `address(0)`, the contract will revert with an
+     * {ERC2771ForwarderMismatchedValue} error.
+     *
+     * NOTE: The `atomic` flag guarantees an all-or-nothing behavior only for the first level forwarded
+     * calls. In case a call is forwarded to another contract, it may revert without the top-level call
+     * reverting.
+     */
+    function executeBatch(
+        ForwardRequestData[] calldata requests,
+        bool atomic,
+        address payable refundReceiver
+    ) public payable virtual {
+        uint256 requestsValue;
+        uint256 spentValue;
+
+        for (uint256 i; i < requests.length; ++i) {
+            requestsValue += requests[i].value;
+            bool success = _execute(requests[i], _useNonce(requests[i].from), atomic);
```

Comment:
Here, if atomic is false:
- we first use the nonce
- we go inside the `_execute` function
  - inside `_execute`, `requireValidRequest` is false, so we don't check the "liveness" of the request.
    if the request is not live we do nothing
    
This means that we used the nonce of `requests[i].from` even though the request might not have been signed. It feels like this allows anyone to invalidate anyone elses nonce. This, combined with frontrunning is really bad. You could DOS someone, of worst. 

Lets say Alice wants to do operations A, B, C and D. I like B and D (they are to my advantage) but not A and C.
I could forge two invalid requests, with Alice as the from, that I would call A2 and C2. I then try to batch A2, B, C2, D with atomic to set to false. B and D will be executed, while A2 and C2 will increase the nonce without doing anything.

Reply:
I'm proposing this fix: https://github.com/Amxx/openzeppelin-contracts/commit/b2c8ec406333b2fb274e7995b5f28d814183506c

Reply:
We ended up removing the nonce for optimizing `calldata`. It's not worth keeping 2 SLOADS when we can add the nonce to `calldata` and validate. I'll revert back to those changes.

Reply:
Comparing gas costs we see that 1 redundant SLOAD costs 100 gas, whereas adding nonce in calldata costs 150 gas. In rollups that post calldata to L1 the different will likely be greater in favor of using storage.